In [46]:
import numpy
import statistics

Với 3 mẫu dữ liệu trên, ta tìm cách sắp chúng lại thành ma trận có kích thước m x N (2 dòng, 3 cột). Để làm điều đó dễ dàng, ta sắp chúng thành ma trận có kích thước N x m trước, sau đó chuyển vị để được ma trận m x N.

In [88]:
# Cho 3 mẫu dữ liệu
A = numpy.array([-2,-2])
B = numpy.array([-1,4])
C = numpy.array([2,3])

Ma trận N x m:

In [89]:
# Hình thành ma trận
Matrix = numpy.array([A,B,C])
print(Matrix)

[[-2 -2]
 [-1  4]
 [ 2  3]]


Chuyển vị ma trận N x m thành ma trận m x N, đặt tên là K:

In [90]:
K = numpy.transpose(Matrix)
print(K)

[[-2 -1  2]
 [-2  4  3]]


Ma trận K có kích thước m x N (2x3) với m là số chiều của mẫu dữ liệu, N là số mẫu dữ liệu.

**II. Tính mẫu trung bình**

Do mỗi điểm của chúng ta nằm trên không gian 2 chiều (m=2), do đó khi ta tìm trọng tâm M (giá trị trung bình của N mẫu trên không gian m chiều, tạm gọi nó là mẫu trung bình) thì trọng tâm này cũng là 2 chiều

In [144]:
M = []
for i in range(0,len(K)):
    M.append(numpy.mean(K[i]))

M = numpy.array(M)
print(M)



[-0.33333333  1.66666667]


**III. Trừ mỗi mẫu với giá trị trung bình để tìm độ lệch**

Ta nhân bản ma trận M thành tương ứng N mẫu được ma trận M2:

In [177]:
M2 = numpy.tile(M,(len(Matrix),1))
M2 = M2.T

Sau đó trừ từng phân tử của ma trận M với từng phân tử tương ứng của ma trận M

In [180]:
D = K - M2
D

array([[-1.66666667, -0.66666667,  2.33333333],
       [-3.66666667,  2.33333333,  1.33333333]])

**IV. Tính ma trận hiệp phương sai m x m**

Ma trận hiệp phương sai C:

In [184]:
N = len(Matrix)
N

3

In [193]:
C = (1/(N-1))*D
numpy.dot(C,numpy.transpose(D))

array([[ 4.33333333,  3.83333333],
       [ 3.83333333, 10.33333333]])

**V. Nhận xét ma trận hiệp phương sai (covariance matrix)**

c11 và c22 lần lượt là phương sai (variance) của trục X và trục Y. Nói cách khác, các phần tử trên đường chéo chính của ma trận hiệp phương sai là các phương sai của các mẫu dữ liệu.

Ma trận hiệp phương sai có tính chất đối xứng qua đường chéo chính
- c11 = 4.33 
- c22 = 10.33

**Kiểm chứng kết quả tính toán ma trận hiệp phương sai bằng thư viện có sẵn của python (covariance matrix)**


In [194]:
print(f'cov: {numpy.cov(Matrix.T)}')

cov: [[ 4.33333333  3.83333333]
 [ 3.83333333 10.33333333]]


# Nguồn tài liệu tham khảo:

https://minhng.info/toan-hoc/ma-tran-hiep-phuong-sai-covariance-matrix.html